add markdown support

jorolf opened the pull request:
if you **really** want to emphasize _something_

jorolf committed:
add markdown support

jorolf committed:
fix appveyor warnings

Nabile-Rahmani commented on osu.Game/Overlays/Chat/ChatLine.cs:
```diff
@@ -112,16 +115,33 @@ public ChatLine(Message message)
+                string[] textArray = text.Split(new string[] { "*" }, StringSplitOptions.None);
+                for (int i = 0; i < textArray.Length; i++)
+                {
+                    if (i != 0) //we shouldn't switch when the i is zero because then there wasn't a asterisk yet and we don't want to switch at the end because then there could've just been two asterisks
```
Can't you swap out that toggle for `italic = i % 2 != 0` ?

Jorolf replied:
that wouldn't support random bold text in an italic one

Tom94 commented:
First, this needs to support escaping via \. Smilies like *_* are impossible otherwise.
Next, if we go this way, then _ should be permitted for italic text, too, as it is very common in chat applications. Lastly, as far as I can tell the current implementation switches to italic if there is just a single * in a sentence. That should not happen either.

jorolf committed:
rework logic

jorolf committed:
make CI happy

Tom94 committed:
Nicer parsing for markdown (currently only supports bold and italic)

Tom94 committed:
Better name for parse method

Tom94 committed:
Fix incorrect escaping behavior

Previously, escaping was not correct. Now, the escape character
correctly only escapes a single character of the delimiter.

Tom94 committed:
Add private keyword

huoyaoyuan commented on osu.Game/Overlays/Chat/ChatLine.cs:
```diff
@@ -112,16 +116,116 @@ public ChatLine(Message message)
+            markers.Add(new SplitMarker { Index = toParse.Length, Length = 0, Type = SplitType.None });
+
+            // Sort markers from earliest to latest
+            markers.Sort((a, b) => a.Index.CompareTo(b.Index));
```
Make `SplitMarker : IComparable<SplitMarker>`, and then default sort method can be used.

Tom94 replied:
Why overengineer if this is perfectly sufficient? I prefer to leave the struct be just a dumb data container unless the abstractions are actually useful in multiple places.

Tom94 committed:
Make split markers be based on flags

This avoids the need for explicit booleans for each split type and
elegantly replaces the switch with a xor.

Tom94 committed:
Simplify code a bit

Tom94 committed:
Add flags attribute

Tom94 committed:
Stop CI from complaining about modified closure

jorolf committed:
Merge branch 'master' into markdown

DrabWeb commented:
I think this should be moved to the framework with `TextFlowContainer` (that's why it was originally called `Paragraph`, this was planned to be added).

peppy commented:
I agree we need this in the framework. I'm not sure it makes sense to implement it just for ChatLine here.

Tom94 commented:
If we put it into framework, then it should be a wrapper around the `TextFlowContainer` we have right now. Basically, taking markdown-formatted text and outputting a `TextFlowContainer` which has the correct pieces.

peppy commented:
Any chance you can reimplement this in the framework as discussed @jorolf? We definitely want to use it in more places than just `ChatLine`.

peppy reviewed (changes requested):
Should be added to framework so we can reuse this across multiple use cases.

jorolf commented:
working on a framework implementation
